File: go/instruction_serialiser/add_node.go
package instruction_serialiser

func (n *AddNode) calculate(parameters map[string]interface{}) (ArithmeticType, error) {
	var leftVal ArithmeticType
	{
		var err error
		if leftVal, err = n.GetLeftChild().Calculate(parameters); err != nil {
			return 0, err
		}
	}
	var rightVal ArithmeticType
	{
		var err error
		if rightVal, err = n.GetRightChild().Calculate(parameters); err != nil {
			return 0, err
		}
	}
	return leftVal + rightVal, nil
}

func (a *ArithmeticResultNodeWrapper) Add(node *ArithmeticResultNodeWrapper) *ArithmeticResultNodeWrapper {
	return &ArithmeticResultNodeWrapper{
		Node: &ArithmeticResultNodeWrapper_AddNode{
			AddNode: &AddNode{
				LeftChild:  a,
				RightChild: node,
			},
		},
	}
}

File: rust/src/add_node.rs
use std::collections::HashMap;
use std::ops::{Add, AddAssign};
use crate::arithmetic_type::ArithmeticType;
use crate::generic_node::GenericNode;
use crate::calculation_error::CalculationError;
use crate::instruction_serialiser::{AddNode, ArithmeticResultNodeWrapper};
use crate::instruction_serialiser::arithmetic_result_node_wrapper::Node;
use crate::node_wrapper::NodeWrapper;
use crate::parameter::Parameter;

impl GenericNode<ArithmeticType> for AddNode {
    fn calculate(&self, parameters: &HashMap<&str, Parameter>) -> Result<ArithmeticType, CalculationError> {
        let left_node = self.left_child.as_ref().ok_or(
            CalculationError::new("The left child of an add node must be present to perform a calculation")
        )?;
        let right_node = self.right_child.as_ref().ok_or(
            CalculationError::new("The right child of an add node must be present to perform a calculation")
        )?;
        Ok(left_node.calculate(Some(parameters))? + right_node.calculate(Some(parameters))?)
    }
}

type Arnw = ArithmeticResultNodeWrapper;

impl Add for Arnw {
    type Output = Self;

    fn add(self, node: Self) -> Self::Output {
        Self{
            node: Some(
                Node::AddNode(
                    Box::new(
                        AddNode{
                            left_child: Some(Box::new(self)),
                            right_child: Some(Box::new(node))
                        }
                    )
                )
            )
        }
    }
}

impl AddAssign for Arnw {
    fn add_assign(&mut self, node: Self) {
        *self = self.clone().add(node);
    }
}
